src/cdomains/apron/elinaImplementation.elina.ml

```diff
@@ -0,0 +1,154 @@
+    if Mpqf.sgn inf1 = 1 && Mpqf.sgn sup1 = 1 then
+      mts zero sup2mo
+    (* If dividend is always negative the range is [-S2+1;0] or possibly a smaller range *)
+    else if Mpqf.sgn inf1 = -1 && Mpqf.sgn sup1 = -1 then
+      mts (Mpqf.neg sup2mo) zero
+    (* If dividen could be both negative or positive the range could be [-S2+1;S2-1]*)
+    else
+      mts (Mpqf.neg sup2mo) sup2mo
+    (* This fails to consider many subcases such as the fact that [1;2]%[3;4]
+    would be in range [1;2] not [1;3], etc. Also I'm not even sure I understand the negative
+    result thing correctly*)
+
+  (* Binop/Pow case *)
+  let rec bound_texpr_pow_rec a b = 
+    let one = Mpqf.of_int 1 in
+    if Mpqf.sgn b = 1 then 
+      Mpqf.mul a (bound_texpr_pow_rec a (Mpqf.sub b one))
+    else if Mpqf.sgn b = -1 then
+      Mpqf.div one (bound_texpr_pow_rec a (Mpqf.neg b))
+    else
+      one
+    
+  let bound_texpr_pow i1 i2 s1 s2 =
+    (* TO DO *)
+    mts (bound_texpr_pow_rec (stm i1) (stm i2)) (bound_texpr_pow_rec (stm s1) (stm s2))
+
+  (* Unop/Sqrt case *)
+  let bound_texpr_sqrt inf sup =
+    (* Very poor sqrt approximation *)
+    mts (Mpqf.of_int 0) (Mpqf.add (Mpqf.div (stm sup) (Mpqf.of_int 2)) (Mpqf.of_int 1))
+
+  (* Workaround for Abstract1.bound_texpr *)
+  (* This creates an approximation based on the bounds
+  of the component variables which is not as precise
+  as the actual bound_texpr could be
+  *)
+  let rec bound_texpr_alt mgr d exprt1 =
+    match exprt1 with
+    (* Constant *)
+    | Texpr1.Cst cst ->
+      (match cst with
+      | Coeff.Scalar coeff -> Interval.of_scalar coeff coeff
+      | Coeff.Interval intv -> intv
+      )
+    (* Variable *)
+    | Texpr1.Var var -> Abstract1.bound_variable mgr d var
+    (* Unary *)
+    | Texpr1.Unop (unop,expr,typ,round) ->
+      let bounds = bound_texpr_alt mgr d expr in
+      (match unop with
+      | Texpr1.Neg -> Interval.of_scalar (Scalar.neg bounds.inf) (Scalar.neg bounds.sup)
+      | Texpr1.Cast -> bounds (* Unsure? *)
+      | Texpr1.Sqrt -> bound_texpr_sqrt bounds.inf bounds.inf
+      )
+    (* Binary *)
+    | Texpr1.Binop (binop,expr1,expr2,typ,round) ->
+      let bounds1 = bound_texpr_alt mgr d expr1 in
+      let bounds2 = bound_texpr_alt mgr d expr2 in
+      (match binop with
+      | Texpr1.Add -> mts (Mpqf.add (stm bounds1.inf) (stm bounds2.inf)) (Mpqf.add (stm bounds1.sup) (stm bounds2.sup))
+      | Texpr1.Sub -> mts (Mpqf.sub (stm bounds1.inf) (stm bounds2.sup)) (Mpqf.sub (stm bounds1.sup) (stm bounds2.inf))
+      | Texpr1.Mul -> mts (Mpqf.mul (stm bounds1.inf) (stm bounds2.inf)) (Mpqf.mul (stm bounds1.sup) (stm bounds2.sup))
+      | Texpr1.Div -> mts (Mpqf.div (stm bounds1.inf) (stm bounds2.sup)) (Mpqf.div (stm bounds1.sup) (stm bounds2.inf))
+      | Texpr1.Mod -> bound_texpr_mod bounds1.inf bounds2.inf bounds1.sup bounds2.sup
+      | Texpr1.Pow -> bound_texpr_pow bounds1.inf bounds2.inf bounds1.sup bounds2.sup

```

Comment:
What happens in the operations yield results that are out of bounds for the data type? Does one need to check whether the bounds are within the range of the data type and otherwise do some overflow handling? Or is this check for being in type bounds done somewhere else?

Reply:
I believe this is handled by the overflow_handling function from shared functions where this function is called from where it checks whether the value is in bounds of the datatype with "Z.compare type_min min <= 0 && Z.compare max type_max <= 0"